# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: accessible/ipc/RemoteAccessibleBase.h
# Commit: 3dd1b7e892e0
# Full Hash: 3dd1b7e892e013f0909e298fff25d5c4602a27e4
# Author: James Teh <jteh@mozilla.com>
# Date: 2022-05-11 09:42:52
# Description:
#   Bug 1768391: Invalidate the remote HyperText offsets cache when we mutate the tree. r=eeejay
#   
#   Previously, we invalidated this cache when the reorder event arrived.
#   Because the mutation and the reorder event happen in separate IPDL calls, it's possible for a client call to arrive between them.
#   If that client call queried HyperText offsets, this could result in returning incorrect information to the client or even a parent process crash.
# ==============================================================================

diff -r 7fd565174880 -r 3dd1b7e892e0 accessible/ipc/RemoteAccessibleBase.h
--- a/accessible/ipc/RemoteAccessibleBase.h	Tue May 10 22:53:56 2022 +0000
+++ b/accessible/ipc/RemoteAccessibleBase.h	Tue May 10 22:56:57 2022 +0000
@@ -35,6 +35,9 @@
 
   void AddChildAt(uint32_t aIdx, Derived* aChild) {
     mChildren.InsertElementAt(aIdx, aChild);
+    if (IsHyperText()) {
+      InvalidateCachedHyperTextOffsets();
+    }
   }
 
   virtual uint32_t ChildCount() const override { return mChildren.Length(); }
@@ -114,7 +117,12 @@
   /**
    * Remove The given child.
    */
-  void RemoveChild(Derived* aChild) { mChildren.RemoveElement(aChild); }
+  void RemoveChild(Derived* aChild) {
+    mChildren.RemoveElement(aChild);
+    if (IsHyperText()) {
+      InvalidateCachedHyperTextOffsets();
+    }
+  }
 
   /**
    * Return the proxy for the parent of the wrapped accessible.
@@ -245,6 +253,12 @@
         mCachedFields = new AccAttributes();
       }
       mCachedFields->Update(aFields);
+      if (IsTextLeaf()) {
+        Derived* parent = RemoteParent();
+        if (parent && parent->IsHyperText()) {
+          parent->InvalidateCachedHyperTextOffsets();
+        }
+      }
     }
   }
 
